sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisCursor.java

```diff
@@ -0,0 +1,149 @@
+import java.util.Objects;
+import javax.annotation.Nonnull;
+import org.apache.beam.sdk.coders.BigEndianLongCoder;
+import org.apache.beam.sdk.io.range.ByteKey;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class RedisCursor implements Comparable<RedisCursor>, Serializable {
+
+  public static final RedisCursor ZERO_CURSOR = RedisCursor.of("0", 8);
+
+  private final String cursor;
+  private final ByteKey byteCursor;
+  private final long dbSize;
+  private final int nBits;
+
+  public static RedisCursor of(String cursor, long dbSize) {
+    return new RedisCursor(cursor, dbSize);
+  }
+
+  public static RedisCursor of(ByteKey byteCursor, long dbSize) {
+    return new RedisCursor(byteCursor, dbSize);
+  }
+
+  private RedisCursor(ByteKey byteCursor, long dbSize) {
+    this.byteCursor = byteCursor;
+    this.dbSize = dbSize;
+    this.nBits = getTablePow(dbSize);
+    this.cursor = byteKeyToString(byteCursor, nBits);
+  }
+
+  private RedisCursor(String cursor, long dbSize) {
+    this.cursor = cursor;
+    this.dbSize = dbSize;
+    this.nBits = getTablePow(dbSize);
+    this.byteCursor = stringCursorToByteKey(cursor, this.nBits);
+  }
+
+  /**
+   * {@link RedisCursor} implements {@link Comparable Comparable&lt;RedisCursor&gt;} by transforming
+   * the cursors to an index of the Redis table.
+   */
+  @Override
+  public int compareTo(@Nonnull RedisCursor other) {
+    checkNotNull(other, "other");
+    return Long.compare(Long.parseLong(cursor), Long.parseLong(other.cursor));
+  }
+
+  @Override
+  public boolean equals(@Nullable Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    RedisCursor that = (RedisCursor) o;
+    return dbSize == that.dbSize
+        && nBits == that.nBits
+        && Objects.equals(cursor, that.cursor)
+        && Objects.equals(byteCursor, that.byteCursor);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(cursor, byteCursor, dbSize, nBits);
+  }
+
+  public String getCursor() {
+    return cursor;
+  }
+
+  public ByteKey getByteCursor() {
+    return byteCursor;
+  }
+
+  public long getDbSize() {
+    return dbSize;
+  }
+
+  @VisibleForTesting
+  static ByteKey stringCursorToByteKey(String cursor, int nBits) {
+    long cursorLong = Long.parseLong(cursor);
+    long reversed = shiftBits(cursorLong, nBits);
+    BigEndianLongCoder coder = BigEndianLongCoder.of();
+    ByteArrayOutputStream os = new ByteArrayOutputStream();
+    try {
+      coder.encode(reversed, os);
+    } catch (IOException e) {
+      throw new IllegalArgumentException("invalid redis cursor " + cursor);
+    }
+    byte[] byteArray = os.toByteArray();
+    return ByteKey.copyFrom(byteArray);
+  }
+
+  @VisibleForTesting
+  static long shiftBits(long a, int nBits) {
+    long b = 0;
```

Comment:
Can you simplify this with Long.reverse?

It will also help with readability.

Reply:
+1 - please use Long.reverse(a) , unless you need the `nBits` argument?

Reply:
you can do the following right?

```
nBits = getTablePow(cursor.getDbSize());
Long.reverse(cursorLong) >> (64 - nBits)
```

Reply:
can you please use Long.reverse?

Reply:
I tried changing to    ` long reversed = Long.reverse(cursorLong) >> (64 - nBits);` but the test https://github.com/apache/beam/blob/4adce5979bf7866efa7041f8399703fe31b6053a/sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java#L397` failed with:
`expected:<[1885267]> but was:<[-211885]>`